docs/formats/zjson.md

```diff
@@ -0,0 +1,492 @@
+applies to this value where the definition for the type appears in the
+types field or in a previous occurrence of the types field in the stream,
+* `values` a JSON array of strings and arrays encoded as defined below,
+* `types` a JSON array of types where the types contain "TypeDef" types
+establishing a binding between the names referred to by the `schema` fied.
+
+The schema name provides a mapping to a type so that future values in the stream may
+reference a type by name.  An implementation maintains a table to map schema names
+to types as it decodes values.  The names are scoped to the particular ZJSON
+data stream in which they are embedded and otherwise have no global persistence
+or meaning.
+
+Objects in a ZJSON stream have the following JSON structure:
+```
+{
+  "schema": <id>,
+  "values": [ <val>, ... [ <val>, ... ] ... ]
+  "types": [ <type>, <type>, ... ]
+}
+```
+
+### Type Encoding
+
+The type format follows the terminology in the ZSON spec, where primitive types
+represent concrete values like strings, integers, times, and so forth, while
+complex types are composed of primitive types and/or other complex types, e.g.,
+records, sets, arrays, and unions.
+
+The ZJSON type encoding for a primitive type is simply its ZSON string name,
+e.g., "int32" or "string".  Complex types are structured and their
+mapping onto JSON depends on the type.  For example,
+the Zed type `{s:string,x:int32}` has this ZJSON format:
+```
+{
+  "kind": "record",
+  "fields": [
+    {
+      "name": "s",
+      "type": {
+        "kind": "primitive",
+        "name": "string"
+      }
+    },
+    {
+      "name": "x",
+      "type": {
+        "kind": "primitive",
+        "name": "int64"
+      }
+    }
+  ]
+}
+```
+A type string may also contain a type name previously defined by a type definition.
+
+#### Record Type
+
+More formally, a Zed record type is a JSON object of the form
+```
+{
+  "kind": "record",
+  "fields": [ <field>, <field>, ... ]
+}
+```
+where each of the fields has the form
+```
+{
+  "name": <name>,
+  "type": <type>,
+}
+```
+and `<name>` is a string defining the column name and `<type>` is a
+recursively encoded type.
+
+#### Array Type
+
+A Zed array type is defined by a JSON object having the form
+```
+{
+  "kind": "array",
+  "type": <type>
+}
+```
+where `<type>` is a recursively encoded type.
+
+#### Set Type
+
+A Zed set type is defined by a JSON object having the form
+```
+{
+  "kind": "set",
+  "type": <type>
+}
+```
+where `<type>` is a recursively encoded type.
+
+#### Union type
+
+A Zed union type is defined by a JSON object having the form
+```
+{
+  "kind": "union",
+  "types": [ <type>, <type>, ... ]
+}
+```
+where the list of types comprise the types of the union and
+and each `<type>`is a recursively encoded type.
+
+#### Enum Type
+
+A Zed enum type is a JSON object of the form
+```
+{
+  "kind": "enum",
+  "symbols": [ <string>, <string>, ... ]
+}
+```
+
```

Comment:
Don't forget maps!
~~~
#### Map Type

A Zed map type is defined by a JSON object of the form
```
{
  "kind": "map",
  "key_type": <type>,
  "val_type": <type>
}
```

~~~